cmd/kubeadm/app/phases/certs/certs.go

```diff
@@ -120,6 +122,105 @@ func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProx
 	return frontProxyClientCert, frontProxyClientKey, nil
 }
 
+// UsingExternalCA determines whether the user is relying on an external CA.  We currently implicitly determine this is the case when the CA Cert
+// is present but the CA Key is not. This allows us to, e.g., skip generating certs or not start the csr signing controller.
+func UsingExternalCA(cfg *kubeadmapi.MasterConfiguration) bool {
+
+	if !ValidateCACert(cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, "CA") {
+		return false
+	}
+
+	// Check CA Key is not present
+	caKeyPath := filepath.Join(cfg.CertificatesDir, kubeadmconstants.CAKeyName)
+	if _, err := os.Stat(caKeyPath); !os.IsNotExist(err) {
+		return false
+	}
+
+	if !ValidateSignedCert(cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, kubeadmconstants.APIServerCertAndKeyBaseName, "API server") {
+		return false
+	}
+
+	if !ValidateSignedCert(cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName, "API server kubelet client") {
+		return false
+	}
+
+	if !ValidatePrivateKey(cfg.CertificatesDir, kubeadmconstants.ServiceAccountKeyBaseName, "service account") {
+		return false
+	}
+
+	if !ValidateCACertAndKey(cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertAndKeyBaseName, "front-proxy CA") {
+		return false
+	}
+
+	if !ValidateSignedCert(cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertAndKeyBaseName, kubeadmconstants.FrontProxyClientCertAndKeyBaseName, "front-proxy client") {
+		return false
+	}
+
+	return true
+}
+
+func ValidateCACert(pkiDir string, baseName string, UXName string) bool {
+	// Check CA Cert
+	caCert, err := pkiutil.TryLoadCertFromDisk(pkiDir, baseName)
+	if err != nil {
+		fmt.Printf("failure loading certificate for %s: %v", UXName, err)
+		return false
+	}
+
+	// Check if cert is a CA
+	if !caCert.IsCA {
+		fmt.Printf("certificate %s is not a CA", UXName)
+		return false
+	}
+	return true
+}
+
+func ValidateCACertAndKey(pkiDir string, baseName string, UXName string) bool {
```

Comment:
Table driven tests are common practice, please write such ones.
Nearly all unit tests in `cmd/kubeadm` are table driven